### PR TITLE
[webdriverio] Nullable Hooks, fix reporterOptions, add `deprecationWarnings`, add `browserstackLocal`

### PR DESCRIPTION
**Pull request by phil-lgr**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: 

`browserstackLocal` added as boolean: https://github.com/webdriverio/webdriverio/blob/82dd1cabb07bb1c999ed28372d37aa0c8bd1b9d2/docs/guide/services/browserstack.md#configuration

`reporterOptions` should be more flexible to allow reporter custom config: https://github.com/webdriverio/wdio-junit-reporter#configuration

`deprecationWarnings` was introduced in 4.10.2: https://github.com/webdriverio/webdriverio/commits/a8be2255b3700999fe17d3bfde5b56e68d3cd15a/lib/helpers/deprecationWarning.js

I also think we could set all the hook as nullable:

```
before?<T>(
            capabilities: DesiredCapabilities,
            specs: string[]
        ): Promise<T> & undefined;
```

One can choose not to specify a `before` hook in the config file. 

- [X] Increase the version number in the header if appropriate.

I'm running webdriverio 4.10.2 against @types/webdriverio 4.8.0. With those changes it looks good.

```
// Type definitions for WebdriverIO 4.10.2
// Project: http://www.webdriver.io/
// Definitions by: Nick Malaguti <https://github.com/nmalaguti>
//                 Tim Brust <https://github.com/timbru31>
//                 Fredrik Smedberg <https://github.com/fsmedberg-tc>
//                 Tanvir ul Islam <https://github.com/tanvirislam06>
//                 Phil Leger <https://github.com/phil-lgr>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

/// <reference types="node"/>
```

is 4.10.2 good or should it be 4.10?

### CODE REVIEW

**Comment by phil-lgr** [comment]
I have a question for the authors at the same time... I am not sure I understand why all hooks methods are defined as generic:

image

It makes this really hard to `implement` the Config:

```
export const wdioCommonConfig: WebdriverIO.Config
```

Possible solution, make the Hooks interface generic:
```
interface Hooks<T> {
        onError?(error: Error): Promise<T> & undefined;

        onPrepare?(
            config: Options,
            capabilities: DesiredCapabilities
        ): Promise<T> & undefined;

        onComplete?(exitCode: number): Promise<T> & undefined;
```

Thoughts?

**Comment by timbru31** [comment]
@phil-lgr - care to upgrade to v4.11? Just saw that a new version was released. 👍

File: types/webdriverio/index.d.ts [hunk 1]
```diff
@@ -1,9 +1,10 @@
-// Type definitions for WebdriverIO 4.8
+// Type definitions for WebdriverIO 4.10
```
Comment (timbru31) [inline types/webdriverio/index.d.ts]: just for reference: semver minor is the correct level. patch versions are omitted.

**Comment by phil-lgr** [comment]
@timbru31 I will submit a new PR soon, I'm doing a full setup with wdio using TypeScript so I am seeing other definitions that needs to be updated - I will check 4.11 at the same time.

To give you an idea:

image

**Pull request merged**